backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CpuPinningPolicyUnit.java

```diff
@@ -58,78 +60,85 @@ public CpuPinningPolicyUnit(PolicyUnit policyUnit,
-                                vmToPendingDedicatedCpuPinnings,
-                                vm,
-                                host.getId());
-                if (isDedicatedCpuPinningPossibleAtHost) {
-                    candidates.add(host);
-                } else {
-                    messages.addMessage(host.getId(), EngineMessage.VAR__DETAIL__VM_PINNING_DEDICATED_NOT_FIT.name());
-                    log.debug("Host {} does not satisfy CPU pinning constraints, cannot match virtual topology " +
-                            "with available CPUs.", host.getId());
-                }
             }
-            return candidates;
-        }
-        // shared CPUs
-        if (StringUtils.isEmpty(cpuPinning)) {
-            // For 'Resize and Pin NUMA', on RunVm, we get here before determining the CPU pinning for the VM
-            // so the dynamic CPU pinning is not set and hosts should not be filtered by the logic below.
-            return hosts;
-        }
-        // collect all pinned host cpus and merge them into one set
-        final Set<Integer> pinnedCpus = CpuPinningHelper.getAllPinnedPCpus(cpuPinning);
-        for (final VDS host : hosts) {
-            final Collection<Integer> onlineHostCpus = SlaValidator.getOnlineCpus(host);
-
-            Map<Guid, List<VdsCpuUnit>> vmToPendingDedicatedCpuPinnings =
-                    PendingCpuPinning.collectForHost(getPendingResourceManager(), host.getId());
-            var cpuTopology = resourceManager.getVdsManager(host.getId()).getCpuTopology();
-            vdsCpuUnitPinningHelper.previewPinOfPendingExclusiveCpus(cpuTopology, vmToPendingDedicatedCpuPinnings);
-            final Collection<Integer> dedicatedCpus = cpuTopology.stream().filter(VdsCpuUnit::isExclusive).map(VdsCpuUnit::getCpu).collect(Collectors.toList());
 
-            if (!dedicatedCpus.isEmpty() && vm.getCpuPinningPolicy() == CpuPinningPolicy.RESIZE_AND_PIN_NUMA && pinnedCpus.isEmpty()) {
+            // shared CPUs
+            Set<Integer> pinnedCpus = new LinkedHashSet<>();
+            // collect all pinned host cpus and merge them into one set
+            sharedVms.forEach(vm -> pinnedCpus.addAll(CpuPinningHelper.getAllPinnedPCpus(vm.getCpuPinning())));
+            Optional<VM> resizeWithoutPinning = sharedVms.stream()
+                    // we know based on the above filter that sharedVms contains pinning
+                    .filter(vm -> vm.getCpuPinningPolicy() == CpuPinningPolicy.RESIZE_AND_PIN_NUMA)
+                    .findFirst();
+            final Collection<Integer> onlineHostCpus = SlaValidator.getOnlineCpus(host);
+            final Collection<Integer> dedicatedCpus = cpuTopology.stream()
+                    .filter(VdsCpuUnit::isExclusive)
+                    .map(VdsCpuUnit::getCpu)
+                    .collect(Collectors.toList());
+            if (!dedicatedCpus.isEmpty() && resizeWithoutPinning.isPresent()) {
```

Comment:
shouldn't resizeWithoutPinning reflect on whether there is a vm with resize-and-pin and *withouth* pinning?